AIBDP-9091: Update DOX public client library with Schema API changes

I527266-Kaushiki opened the pull request:
This pull request includes the Schema and Schema Version APIs for Document Information Extraction.
The APIs included are:
  1. Create schema
  2. Get all schemas
  3. Delete specific schema
  4. Create schema version
  5. Get specific schema details
  6. Update schema details
  7. Create schema fields
  8. Activate schema
  9. Deactivate schema
  10. Get schema capabilities
  11. Delete version of schema
  12. Get versions of schema
  13. Get specific version of schema
  14. Along with an additional endpoint with mixture of schema creation and fields creation (Create schema + Create schema fields)

I527266-Kaushiki commented:
Hello @tillgeissler . Can you please review this PR once? Thank you!

tillgeissler commented on sap_business_document_processing/document_information_extraction_client/constants.py:
```diff
@@ -6,17 +6,38 @@
 API_FIELD_RESULTS = 'results'
 API_FIELD_RETURN_NULL = 'returnNullValues'
 API_FIELD_SCHEMA_ID = 'schemaId'
+API_FIELD_SCHEMA_NAME = 'name'
```
This constants might be misleading since it sounds as if the string would be 'schemaName'. Maybe rename to `API_FIELD_NAME`

I527266-Kaushiki replied:
Updated.

tillgeissler reviewed (changes requested):
In general, I am wondering if it is necessary to have most of the functions twice with and without underscore since there is no additional functionality between those. 
Additionally, the library usually returns the json dictionary contained in the response instead of the entire response object, i.e. the functions should return `response.json()`. 
To improve the usability for users, functions 

- should also be named according to the purpose they serve (e.g. "create_schema" instead of "post_schema_configuration")
- should support the user in using the API by providing meaningful parameter names and not let the user compile complex payloads
- should include docstring explaining each parameter and the purpose of the function as well as what is returned. For more complex functions this could include an example usage.

I527266-Kaushiki commented on sap_business_document_processing/common/exceptions.py:
```diff
@@ -28,3 +27,24 @@ class BDPServerException(BDPApiException):
+
+class DOXInternalServerError(DoxApiClientError):
+    def __init__(self, message: str):
+        super().__init__(message=message, status_code=500)
```
Updated.

I527266-Kaushiki commented on sap_business_document_processing/document_information_extraction_client/dox_api_client.py:
```diff
@@ -633,3 +640,247 @@ def post_confirm_document(self, document_id, data_for_retraining=False) -> dict:
+        url = SCHEMAS_CAPABILITIES
+        return self.get(url)
+
+    def delete_schema_versions(self, schema_id, version, client_id):
+        """
+        DELETE /schemas/{schemaId}/versions
+        """
+        response = self._delete_schema_versions(schema_id, version, client_id)
+        return response
+
+    def _delete_schema_versions(self, schema_id, version, client_id):
+        """
+        Delete Versions associated with Schema
+        """
+        url = SCHEMAS_VERSIONS_ENDPOINT.format(schemaId=schema_id) + "?clientId={}".format(client_id)
+        payload = {
+            "version": [
+                version
+            ]
+        }
+        return self.delete(url, json=payload)
```
Hi @tillgeissler . It is possible to delete one version at a time as well as multiple versions at a time if we pass a list of versions; while passing an empty list deletes all the versions, including the original version '1' if its in deactivated state, else throws 400 Bad request telling one or more schema versions are in active state. 

So should I keep one method, changing the parameter name from version to versions, and mentioning in the docstring to pass the versions parameter as a list, even if it has one element or null ?

tillgeissler replied:
I would create a second function that explicitly handles deleting multiple/all versions based on the parameter.

I527266-Kaushiki commented on sap_business_document_processing/document_information_extraction_client/dox_api_client.py:
```diff
@@ -633,3 +640,247 @@ def post_confirm_document(self, document_id, data_for_retraining=False) -> dict:
+            try:
+                with open(file_path, 'w') as file:
+                    json.dump(capabilities, file, indent=4)
+                    file.flush()
```
Hi Till. Basically we didn't want to call the capabilities API for each of the header_field or line_item_field, as it would take more time. That's why we decided to store the capabilities temporarily on the disk. If there is any other way of saving time, can you suggest?

I527266-Kaushiki commented on sap_business_document_processing/document_information_extraction_client/helpers.py:
```diff
@@ -86,3 +93,126 @@ def create_capability_mapping_options(document_type, file_type, header_fields=No
+        API_FIELD_DESCRIPTION: item[API_FIELD_DESCRIPTION],
+        API_FIELD_LABEL: item[API_FIELD_LABEL],
+        API_FIELD_DEFAULT_EXTRACTOR: {
+            API_FIELD_FIELD_NAME: item[API_FIELD_SCHEMA_NAME]
```
Hi Till. For default models, we thought of keeping the same name for field's name and default extractor name, but should we follow something else than this, please suggest.

tillgeissler replied:
Works for me.

I527266-Kaushiki commented on sap_business_document_processing/document_information_extraction_client/helpers.py:
```diff
@@ -86,3 +93,126 @@ def create_capability_mapping_options(document_type, file_type, header_fields=No
+        API_FIELD_SETUP_TYPE: API_FIELD_STATIC,
+        API_FIELD_SETUP_TYPE_VERSION: setup_type_version,
+        API_FIELD_SETUP: setup_fields,
+        API_FIELD_FORMATTING_TYPE: datatype[item[API_FIELD_SCHEMA_NAME]],
```
Done

I527266-Kaushiki commented on sap_business_document_processing/document_information_extraction_client/helpers.py:
```diff
@@ -86,3 +93,126 @@ def create_capability_mapping_options(document_type, file_type, header_fields=No
+            item_payload = generate_item_payload(item, SETUP_TYPE_VERSION_2,
+                                                 {item[API_FIELD_SCHEMA_NAME]: item[API_FIELD_DATATYPE]},
+                                                 setup_fields=setup_fields)
+            item_payload[API_FIELD_DEFAULT_EXTRACTOR] = {}
```
Hi Till. I simplified the LLM and Template model type implementations, but didn't get any wayout for implementing inside the `generate_item_payload` method. Can you suggest something else?

I527266-Kaushiki commented:
Hi @tillgeissler . I have resolved most of the comments you suggested. In few of them, I still have some doubts, so I've kept them open right now. Could you please check them and give review on the updated code? Thanks!

tillgeissler commented on sap_business_document_processing/document_information_extraction_client/constants.py:
```diff
@@ -44,4 +66,44 @@
+
+SCHEMAS_ENDPOINT = '/schemas'
+SCHEMAS_UPDATE_ENDPOINT = SCHEMAS_ENDPOINT + '/{schemaId}'
+SCHEMAS_LIST_ENDPOINT = SCHEMAS_ENDPOINT + '?clientId={}'
```
This constant does not seem to be used anymore, so it can be removed

I527266-Kaushiki replied:
Done

tillgeissler reviewed (changes requested):
Thank you for the changes. 
I am not so sure whether we need the exact status_code checks for every request. The error handling of 4xx and 5xx codes is done within the get/post/delete functions. All other codes are currently accepted as correct. If explicit code checks should be implemented, I'd prefer them to be performed in those lower level function as well. throwing a `BDPApiException` with message "{url} responded with unexpected status code 3xx but expected 2xx", for example.

I527266-Kaushiki commented on sap_business_document_processing/document_information_extraction_client/dox_api_client.py:
```diff
@@ -633,3 +641,405 @@ def post_confirm_document(self, document_id, data_for_retraining=False) -> dict:
+        :return: The API endpoint response as dictionary
+        """
+        payload = {
+            API_FIELD_VALUE: schema_id
```
Done

I527266-Kaushiki commented on sap_business_document_processing/document_information_extraction_client/dox_api_client.py:
```diff
@@ -633,3 +641,405 @@ def post_confirm_document(self, document_id, data_for_retraining=False) -> dict:
+        :return: The API endpoint response as dictionary
+        """
+        payload = {
+            API_FIELD_EXTRACTED_HEADER_FIELDS: header_fields,
+            API_FIELD_EXTRACTED_LINE_ITEM_FIELDS: line_item_fields
```
Added them as optional parameters, as both can be null lists too.

I527266-Kaushiki commented on sap_business_document_processing/document_information_extraction_client/dox_api_client.py:
```diff
@@ -633,3 +641,405 @@ def post_confirm_document(self, document_id, data_for_retraining=False) -> dict:
+        """
+        if model_type not in SUPPORTED_MODEL_TYPES:
+            error_msg = 'Invalid model. Valid models are {}.'.format(str(SUPPORTED_MODEL_TYPES))
+            raise BDPValueError(error_msg)
```
Done

I527266-Kaushiki commented on sap_business_document_processing/document_information_extraction_client/helpers.py:
```diff
@@ -86,3 +94,98 @@ def create_capability_mapping_options(document_type, file_type, header_fields=No
+        if name_to_type_list[field_name] is None:
+            error_msg = 'fieldName not found in defaultExtractor values. Please provide valid fieldName from ' \
+                    '/capabilities API.'
+            raise BDPValueError(error_msg)
```
Done

I527266-Kaushiki commented on sap_business_document_processing/document_information_extraction_client/helpers.py:
```diff
@@ -86,3 +94,98 @@ def create_capability_mapping_options(document_type, file_type, header_fields=No
+        header_name_to_type = create_list_for_header_and_line_items(header_fields, extracted_header_fields)
+        line_name_to_type = create_list_for_header_and_line_items(line_fields, extracted_line_item_fields)
+
+    if model_type == MODEL_TYPE_DEFAULT:
```
Done

I527266-Kaushiki commented on sap_business_document_processing/common/exceptions.py:
```diff
@@ -28,3 +27,13 @@ class BDPServerException(BDPApiException):
+
+class BDPInternalServerError(BDPServerException):
+    def __init__(self, message: str):
+        super().__init__(message=message, status_code=500)
```
Done

I527266-Kaushiki commented on sap_business_document_processing/document_information_extraction_client/helpers.py:
```diff
@@ -1,12 +1,20 @@
+    SETUP_TYPE_VERSION_2, API_FIELD_IS_LINE_ITEM, API_FIELD_TYPE, API_FIELD_PRIORITY, SETUP_TYPE_AUTO, MODEL_TYPE_LLM, \
+    API_FIELD_DATATYPE, MODEL_TYPE_TEMPLATE, SETUP_TYPE_MANUAL, SETUP_TYPE_PRIORITY
+from ..common.exceptions import BDPInternalServerError, BDPValueError
+from .default_extractor_fields import defExtFields
```
Updated the code using lazy loading, and removed this implementation

I527266-Kaushiki commented:
Hi @tillgeissler . I have added lazy loading for getting the response from capabilities API, by just passing the data inside a variable, not storing it inside any json file. Also, I have removed all the status_code checks. Can you please review it? Thanks!

tillgeissler reviewed (approved):
Looks good to me now. Thank you for your work!

I527266-Kaushiki commented on sap_business_document_processing/document_information_extraction_client/dox_api_client.py:
```diff
@@ -633,3 +639,353 @@ def post_confirm_document(self, document_id, data_for_retraining=False) -> dict:
+        url = SCHEMAS_UPDATE_ENDPOINT.format(schemaId=schema_id)
+        return self.put(url, params=params, json=payload)
+
+    def add_schema_version_fields(self, client_id, schema_id, version, header_fields=[], line_item_fields=[]):
```
Okay
